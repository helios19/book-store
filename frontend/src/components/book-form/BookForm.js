import React, {Component} from 'react';
import PropTypes from "prop-types";
class BookForm extends Component{
    constructor(props) {
        super(props);
        this.state = {
            title:'',
            author:'',
            country:'',
            genre:'',
            year:'',
            borrowed: false
        }
        //If you dont use arrow function you will have to manually bind like this
        //If you dont bind you wont be able to access items in the state of this component because it wont be recognised in lifecycle
        //this.onChange = this.onChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }

    onChange =(e) =>{
        //This is what you do for individual target
        //this.setState({name:e.target.value});
        //But if you have plenty
        this.setState({[e.target.name]:e.target.value})
    }

    onSubmit(e){
        e.preventDefault();
        //Copying state object to newBook
        let newBook = this.state;
        this.props.addBook(newBook);
        //Resetting the fields
        this.setState({
            title:'',
            author:'',
            country:'',
            genre:'',
            year:''
        });
    }
    render() {
        return(
            <form onSubmit={this.onSubmit}>
                <div style={{display:'flex' }}>
                    <input
                        type = "text"
                        name = "title"
                        placeholder="Title"
                        style={leftInput}
                        className="form-control input-sm"
                        value={this.state.title}
                        onChange={this.onChange}
                        required
                    />

                    <input
                        type = "text"
                        name = "author"
                        placeholder="Author"
                        style={rightInput}
                        className="form-control input-sm"
                        value={this.state.author}
                        onChange={this.onChange}
                        required
                    />
                </div>
                <br/>
                <div style={{display:'flex', }}>
                    <input
                        type = "text"
                        name = "genre"
                        placeholder="Genre"
                        style={leftInput}
                        className="form-control input-sm"
                        value={this.state.genre}
                        onChange={this.onChange}
                        required
                    />

                    <input
                        type = "text"
                        name = "year"
                        placeholder="Year"
                        style={rightInput}
                        className="form-control input-sm"
                        value={this.state.year}
                        onChange={this.onChange}
                        required
                    />
                </div>
                <br/>
                <div style={{display:'flex' }}>
                    <input
                        type = "text"
                        name = "country"
                        placeholder="Country"
                        style={leftInput}
                        className="form-control input-sm"
                        value={this.state.country}
                        onChange={this.onChange}
                        required
                    />

                    <span style={rightInput}/>
                </div>
                <br/>
                <br/>
                <input
                    type="submit"
                    value="Submit"
                    className="btn btn-primary btn-sm"
                />
                <br/>
                <br/>
            </form>

        // <div className="panel panel-default">
        //     <div className="panel-heading">
        //         <span className="lead">Book Form</span>
        //     </div>
        //
        //     <div className="panel-body">
        //         <div className="formcontainer">
        //             <form onSubmit={this.onSubmit} name="myForm" className="form-horizontal">
        //                 <div className="row">
        //                     <div className="form-group col-md-12">
        //                         <label className="col-md-2 control-label" htmlFor="title">Title</label>
        //                         <div className="col-md-10">
        //                             <input type="text" value={this.state.title} onChange={this.onChange} id="title"
        //                                    className="form-control input-sm"
        //                                    placeholder="Enter book title" required/>
        //                         </div>
        //                     </div>
        //                 </div>
        //
        //                 <div className="row">
        //                     <div className="form-group col-md-12">
        //                         <label className="col-md-2 control-label" htmlFor="author">Author</label>
        //                         <div className="col-md-10">
        //                             <input type="text" value={this.state.author} onChange={this.onChange} id="author"
        //                                    className="form-control input-sm"
        //                                    placeholder="Enter book author" required/>
        //                         </div>
        //                     </div>
        //                 </div>
        //
        //                 <div className="row">
        //                     <div className="form-group col-md-12">
        //                         <label className="col-md-2 control-label" htmlFor="genre">Genre</label>
        //                         <div className="col-md-10">
        //                             <input type="text" value={this.state.genre} onChange={this.onChange} id="genre"
        //                                    className="form-control input-sm"
        //                                    placeholder="Enter book genre" required/>
        //                         </div>
        //                     </div>
        //                 </div>
        //
        //                 <div className="row">
        //                     <div className="form-group col-md-12">
        //                         <label className="col-md-2 control-label" htmlFor="year">Year</label>
        //                         <div className="col-md-10">
        //                             <input type="text" value={this.state.year} onChange={this.onChange} id="year"
        //                                    className="form-control input-sm"
        //                                    placeholder="Enter book year" required/>
        //                         </div>
        //                     </div>
        //                 </div>
        //
        //                 <div className="row">
        //                     <div className="form-group col-md-12">
        //                         <label className="col-md-2 control-label" htmlFor="country">Country</label>
        //                         <div className="col-md-10">
        //                             <input type="text" value={this.state.country} onChange={this.onChange} id="country"
        //                                    className="form-control input-sm"
        //                                    placeholder="Enter book country" required/>
        //                         </div>
        //                     </div>
        //                 </div>
        //
        //                 <div className="row">
        //                     <div className="form-actions floatRight">
        //                         <input type="submit" value="Submit" className="btn btn-primary btn-sm"/>
        //                     </div>
        //                 </div>
        //             </form>
        //         </div>
        //     </div>
        // </div>

    )
    }
}

const leftInput = {
    flex:'5',
    padding:'5px',
    margin:'10px 10px 0px 0px'
}

const rightInput = {
    flex:'5',
    padding:'5px',
    margin:'10px 0px 0px 10px'
}

const rightBtnInput = {
    flex:'5',
    padding:'5px',
    margin:'10px 0px 0px 10px'
}

BookForm.propTyoes = {
    addBook:PropTypes.func.isRequired,
}

export default BookForm;