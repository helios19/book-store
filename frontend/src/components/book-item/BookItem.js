import React, {Component} from 'react';
import PropTypes from "prop-types";
import BookTwoToneIcon from '@material-ui/icons/BookTwoTone';
import DeleteIcon from '@material-ui/icons/Delete';
import IconButton from '@material-ui/core/IconButton';
import Checkbox from '@material-ui/core/Checkbox';
class BookItem extends Component{

    //dynamic Styling
    infoStyle = () =>{
        return {
            backgroundColor: (this.props.book.id %2) === 0? '#c8d4f7cc' :'',
            padding:'10px',
            borderBottom: '1px #ccc dotted',
            display:'flex',
            alignItems:'center',
            justifyContent:'space-between',
            justifyItems: 'flex-start',
        }
    }

    render() {
        //Destructuring instead of using this.props.book.<variable> you can now use name | surname
        const {id} = this.props.book;
        return(

            <tr>
                <td>
                    <div style={iconBook}>
                        <BookTwoToneIcon style={{color:"#138a04", margin:"0px 10px 0px 0px"}}/>
                    </div>
                </td>
                <td>{this.props.book.title}</td>
                <td>{this.props.book.author}</td>
                <td>{this.props.book.country}</td>
                <td>{this.props.book.genre}</td>
                <td>{this.props.book.year}</td>
                <td>
                    <div>
                        <Checkbox
                            checked={this.props.book.borrowed}
                            onChange={(e) =>this.props.bookAvailabilityChange(this.props.book,e)}
                        />
                    </div>
                </td>
                <td>
                    <div style={buttons}>
                        {/*<IconButton color="secondary" onClick={this.props.removeBook.bind(this,id )} >  Because binding in faling in Jest i have used arrow function binding*/}
                        <IconButton color="secondary" onClick={(e) =>this.props.removeBook(id,e)} >
                            <DeleteIcon />
                        </IconButton>
                    </div>
                </td>

            </tr>

        );
    }
}

const iconBook = {
    display: 'flex',
    justifyContent:'space-between',

}

const buttons = {
    display: 'flex'
}
//PropTypes
BookItem.propTypes = {
    book : PropTypes.object.isRequired
}
export  default  BookItem;