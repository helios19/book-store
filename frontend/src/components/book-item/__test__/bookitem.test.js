import React from "react";
import ReactDOM, {unmountComponentAtNode} from "react-dom";
import BookItem from "../BookItem";
import { act } from "@testing-library/react";
import pretty from "pretty";
//For Snapshots Tests
import renderer from "react-test-renderer";
import BookForm from "../../book-form/BookForm";
let container;
beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    unmountComponentAtNode(container)
    document.body.removeChild(container);
    container = null;
});

//
it("Renders Book Info without crashing and Data Fetching Test", async () => {
    const fakeBook = {
        id: '0',
        title:'Title 0',
        author:'Author 0',
        country:'Country 0',
        genre:'Genre 0',
        year:'Year 0'
    };
    /*jest.spyOn(global, "fetch").mockImplementation(() =>
          Promise.resolve({
              json: () => Promise.resolve(fakeBook)
          })
      );*/
    // Simulate an async call
    global.fetch = jest.fn().mockImplementation(() =>
        Promise.resolve({
            json: () => Promise.resolve(fakeBook),
        })
    );

    // Use the asynchronous version of act to apply resolved promises

    await act(async () => {
        ReactDOM.render(<BookItem book={fakeBook} />, container);
    });

    expect(container.querySelectorAll("td")[1].textContent).toBe(fakeBook.title);
    expect(container.textContent).toContain(fakeBook.author);
    expect(container.textContent).toContain(fakeBook.country);
    expect(container.textContent).toContain(fakeBook.genre);
    expect(container.textContent).toContain(fakeBook.year);

    // remove the mock to ensure tests are completely isolated
    global.fetch.mockClear();
    delete global.fetch;
});

it("Snapshot Test with book Prop",  () => {
    const fakeBook = {
        id: '1',
        title:'Title 1',
        author:'Author 1',
        country:'Country 1',
        genre:'Genre 1',
        year:'Year 1'
    };

    // Simulate an async call
    global.fetch = jest.fn().mockImplementation(() =>
        Promise.resolve({
            json: () => Promise.resolve(fakeBook),
        })
    );

    act( () => {
        ReactDOM.render(<BookItem book={fakeBook} />, container);
    });

    expect(container.querySelectorAll("td")[1].textContent).toBe(fakeBook.title);

    expect(container.textContent).toContain(fakeBook.title);
});

it("matches snapshot with Exactly one book callback prop", ()=>{
    const fakeBook = jest.fn();
    const tree = renderer.create(<BookItem book={fakeBook}/>).toJSON();
    expect(tree).toMatchSnapshot();

});