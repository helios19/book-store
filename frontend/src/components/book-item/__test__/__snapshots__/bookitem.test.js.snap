// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot with Exactly one addBook callback prop 1`] = `
<tr>
  <td>
    <div
      style={
        Object {
          "display": "flex",
          "justifyContent": "space-between",
        }
      }
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        style={
          Object {
            "color": "#138a04",
            "margin": "0px 10px 0px 0px",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M13 13l-3-2.25L7 13V4H6v16h12V4h-5z"
          opacity=".3"
        />
        <path
          d="M18 2H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zM9 4h2v5l-1-.75L9 9V4zm9 16H6V4h1v9l3-2.25L13 13V4h5v16z"
        />
      </svg>
    </div>
  </td>
  <td />
  <td />
  <td />
  <td />
  <td />
  <td>
    <div>
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            className="PrivateSwitchBase-input-8"
            data-indeterminate={false}
            onChange={[Function]}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
      </span>
    </div>
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </span>
      </button>
    </div>
  </td>
</tr>
`;

exports[`matches snapshot with Exactly one book callback prop 1`] = `
<tr>
  <td>
    <div
      style={
        Object {
          "display": "flex",
          "justifyContent": "space-between",
        }
      }
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        style={
          Object {
            "color": "#138a04",
            "margin": "0px 10px 0px 0px",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M13 13l-3-2.25L7 13V4H6v16h12V4h-5z"
          opacity=".3"
        />
        <path
          d="M18 2H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zM9 4h2v5l-1-.75L9 9V4zm9 16H6V4h1v9l3-2.25L13 13V4h5v16z"
        />
      </svg>
    </div>
  </td>
  <td />
  <td />
  <td />
  <td />
  <td />
  <td>
    <div>
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            className="PrivateSwitchBase-input-8"
            data-indeterminate={false}
            onChange={[Function]}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
      </span>
    </div>
  </td>
  <td>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </span>
      </button>
    </div>
  </td>
</tr>
`;
